src/librustc_traits/dropck_outlives.rs

```diff
@@ -193,14 +193,42 @@ fn dtorck_constraint_for_ty<'a, 'gcx, 'tcx>(
             .map(|ty| dtorck_constraint_for_ty(tcx, span, for_ty, depth + 1, ty))
             .collect(),
 
-        ty::TyGenerator(def_id, substs, _) => {
-            // Note that the interior types are ignored here.
-            // Any type reachable inside the interior must also be reachable
-            // through the upvars.
-            substs
+        ty::TyGenerator(def_id, substs, _interior) => {
+            // rust-lang/rust#49918: types can be constructed, stored
+            // in the interior, and sit idle when generator yields
+            // (and is subsequently dropped).
+            //
+            // It would be nice to recursively descend into interior
+            // of a generator just like its upvars to resolve this
+            // (since then we could determine what effects dropping
+            // the generator might have by looking at any drop effects
+            // associated with its interior).
+            //
+            // However, the interior's representation uses things like
+            // TyGeneratorWitness that explicitly assume they are not
+            // traversed in such a manner. So instead, we will
+            // simplify things for now by treating all generators as
+            // if they were like ADT's with drop implementations.
+            //
+            // In particular, skipping over `_interior` is safe
+            // because any side-effects from dropping `_interior` can
+            // only take place through references with lifetimes
+            // derived from lifetimes attached to the upvars, and we
+            // *do* traverse the upvars here.
+
+            let new_constraints: Vec<ty::subst::Kind<'tcx>> =
+                substs.upvar_tys(def_id, tcx).map(|t| t.into()).collect();
+
+            let mut constraint = substs
```

Comment:
> https://github.com/rust-lang/rust/pull/45337/files#r145093242
> It might be a better idea to treat generators like trait object and have them always be dtorck. If generators are "supposed to be" captured by an `impl Trait` or trait object, this shouldn't be a problem in practice..

There's no reason to have both `outlives` and dtorck constraints together.

Reply:
Are you saying @arielb1 that we only need to invoke `constraint.outlives.extend(new_contraints)`, and not also include `dtorck_constraint_for_ty`?

Reply:
Sure just return `new_constraints` as outlives constrains, same as trait objects.